feat: add `assert/has-same-constructor`

abhishekblue opened the pull request:
---
type: pre_push_report
description: Results of running various checks prior to pushing changes. report:
  - task: run_javascript_examples status: na
  - task: run_c_examples status: na
  - task: run_cpp_examples status: na
  - task: run_javascript_readme_examples status: na
  - task: run_c_benchmarks status: na
  - task: run_cpp_benchmarks status: na
  - task: run_fortran_benchmarks status: na
  - task: run_javascript_benchmarks status: na
  - task: run_julia_benchmarks status: na
  - task: run_python_benchmarks status: na
  - task: run_r_benchmarks status: na
  - task: run_javascript_tests status: na ---

Resolves #815.

## Description

> What is the purpose of this pull request?

This pull request:

- Adds a fully implemented `hasSameConstructor` utility to the `@stdlib/assert/` namespace to test if two values have the same constructor.

## Related Issues

> Does this pull request have any related issues?

This pull request:

- Resolves #815

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

- [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

abhishekblue committed:
feat(assert): Add @stdlib/assert/has-same-constructor

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

abhishekblue committed:
fix(assert): update README formatting and examples for #815

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

abhishekblue committed:
style(assert): fix lint errors in has-same-constructor

anandkaranubc commented:
/stdlib lint-autofix

stdlib-bot committed:
fix: resolve lint errors

anandkaranubc commented:
@abhishekblue Thanks for working on this issue. That looks like a good starting point.  

The first step here is to ensure that every comment in [this PR](https://github.com/stdlib-js/stdlib/pull/1357) has been addressed in this one. Also, other similar packages in `assert/*` will be a good reference point too.

Can you also change the title to:

feat: add `assert/has-same-constructor`

anandkaranubc commented:
This will also resolve the CI failures you're currently experiencing.

anandkaranubc commented on lib/node_modules/@stdlib/assert/has-same-constructor/test/test.js:
```diff
@@ -0,0 +1,132 @@
+	t.end();
+});
+
+tape( 'compare two null values', function test( t ) {
```
The test descriptions in this file should be more descriptive and follow project's code conventions. Looking at other packages would be useful to fix this.

anandkaranubc commented on lib/node_modules/@stdlib/assert/has-same-constructor/lib/main.js:
```diff
@@ -0,0 +1,60 @@
+* var x = [];
+* var y = [];
+* var bool = hasSameConstructor( x, y );
+* // returns true
+*
+*/
+
+function hasSameConstructor( x, y ) {
```
```suggestion
* // returns true
*/
function hasSameConstructor( x, y ) {
```

anandkaranubc replied:
Make sure you are following correct spacing conventions throughout this PR.

abhishekblue committed:
made suggested changes in example.js and  main.js

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

abhishekblue committed:
adjust formatting

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: passed
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: failed
---

abhishekblue commented:
@Neerajpathak07 @anandkaranubc need guidance on how can I pass these two checks. I am unable to resolve these. 
Thanks!

abhishekblue committed:
fix(assert): resolve README lint errors and improve formatting

abhishekblue committed:
adjust main.js

abhishekblue commented:
@Neerajpathak07 @anandkaranubc, I’ve been working on this PR but I’m struggling to pass the tests despite multiple attempts. Could you please provide some guidance or point me to resources that might help me resolve these issues? I’d really appreciate your support. Thanks!

kgryte commented on lib/node_modules/@stdlib/assert/package.json:
```diff
@@ -30,8 +30,10 @@
   },
-  "dependencies": {},
-  "devDependencies": {},
+  "dependencies": {
```
In absolutely no circumstances should you be editing this file in this PR. These dependencies should be removed.

abhishekblue replied:
I must've accidentally edited this file. Apologies for this!

kgryte commented on lib/node_modules/@stdlib/assert/has-same-constructor/test/test.js:
```diff
@@ -0,0 +1,76 @@
+});
+
+tape( 'the function returns `true` if provided two arguments which have the same constructor', function test ( t ) {
+    var values;
```
Please make sure you have installed and setup EditorConfig. You have mixed tabs and spaces throughout this file. You can find more instructions in our contributing and development guides.

kgryte commented on lib/node_modules/@stdlib/assert/has-same-constructor/test/test.js:
```diff
@@ -0,0 +1,76 @@
+    var bool;
+    var i;
+
+    values =[
```
You need to rethink these tests. Testing built-in primitives isn't particularly interesting. The entire purpose of this API is to determine whether two class instances share the same original constructor.

abhishekblue replied:
@kgryte, thank you for your feedback. To confirm, should I create custom classes and their constructors and run tests specifically on those constructors to determine whether class instances share the same original constructor? Just want to be sure before proceeding further.

kgryte commented on lib/node_modules/@stdlib/assert/has-same-constructor/lib/main.js:
```diff
@@ -0,0 +1,52 @@
+* // returns true
+*/
+function hasSameConstructor( x, y ) {
+    if ( x === null || y === null ) {
```
What about `undefined`?
